dit/npscalardist.py

```diff
@@ -489,50 +492,231 @@ def from_distribution(cls, dist, base=None, prng=None, extract=True):
+        op : function
+            The comparison or operator to implement.
+        msg : str
+            The error message to raise if the objects are not comparible.
+        """
+        if isinstance(other, ScalarDistribution):
+            from .distconst import _combine_scalar_dists
+            return _combine_scalar_dists(self, other, op)
 
-        The other distribution is assumed to have the same meta information
-        and sample space.
+        elif isinstance(other, numbers.Number):
+            from .distconst import modify_outcomes
+            d = modify_outcomes(self, lambda x: op(x, other))
+            return d
 
-        """
-        # Copy to make sure we don't lose precision when converting.
-        d2 = other.copy(base=self.get_base())
+        else:
+            raise NotImplementedError(msg.format(type(self), type(other)))
 
-        # If self is dense, the result will be dense.
-        # If self is sparse, the result will be sparse.
-        d = self.copy()
-        for outcome, prob in d2.zipped():
-            d[outcome] = d.ops.add(d[outcome], prob)
 
-        return d
+    def __add__(self, other):
+        """
+        Addition of ScalarDistribution with either a scalar or another
+        Scalardistribution.
+        """
+        op = lambda x, y: x + y
+        msg = "Cannot add types {0} and {1}"
+        return self.__meta_magic(other, op, msg)
 
     def __radd__(self, other):
         return self.__add__(other)
 
+    def __sub__(self, other):
+        """
+        Subtraction of ScalarDistribution with either a scalar or another
+        Scalardistribution.
+        """
+        op = lambda x, y: x - y
+        msg = "Cannot subtract types {0} and {1}"
+        return self.__meta_magic(other, op, msg)
+
+    def __rsub__(self, other):
+        """
+        Subtraction of ScalarDistribution from either a scalar or another
+        Scalardistribution.
+        """
+        op = lambda x, y: y - x
+        msg = "Cannot subtract types {1} and {0}"
+        return self.__meta_magic(other, op, msg)
+
     def __mul__(self, other):
         """
-        Scalar multiplication on distributions.
+        Multiplication of ScalarDistribution with either a scalar or another
+        ScalarDistribution.
+        """
+        op = lambda x, y: x * y
+        msg = "Cannot multiply types {0} and {1}"
+        return self.__meta_magic(other, op, msg)
 
-        `other` is assumed to have the same base as `self`.
+    def __rmul__(self, other):
+        return self.__mul__(other)
 
-        The appropriate operation is performed assuming that the scalar
-        multiple is of the same base as `self`. If vanilla scalar
-        multiplication is desired, perform the operation directly on
-        `self.pmf`.
+    def __div__(self, other):
+        """
+        Division of ScalarDistribution with either a scalar or another
+        Scalardistribution.
+        """
+        op = lambda x, y: x / y
```

Comment:
Just seeking confirmation: The intention here is that this should act like floor division if the user has not imported from `__future__`?